panel/io/cache.py

```diff
@@ -305,11 +308,46 @@ def compute_hash(func, hash_funcs, args, kwargs):
         _HASH_MAP[key] = hash_value
     return hash_value
 
+P = ParamSpec("P")
+R = TypeVar("R")
+T = TypeVar("T")
+
+@overload
+def cache(
+    func: Literal[None] = ...,
+    hash_funcs: dict[type[Any], Callable[[Any], bytes]] | None = ...,
+    max_items: int | None = ...,
+    policy: Literal['FIFO', 'LRU', 'LFU'] = ...,
+    ttl: float | None = ...,
+    to_disk: bool = ...,
+    cache_path: str = ...,
+    per_session: bool = ...,
+) -> Callable[[Callable[P, R]], Callable[P, R]]:
```

Comment:
I'm not an expert on type check, so I would like to learn why this is needed. Can you please explain?

Reply:
`cache` has situational behavior depending on the arguments, it can either be a first-order decorator or a second-order decorator.

If you use it as `@pn.cache(ttl=60)`, then you are effectively using as a function that returns a decorator (`pn.cache(ttl=60)(my_func)`) or a second-order decorator. That is one of the overloads.

If you use it as `@st.cache`. you are using it as a first-order decorator (`pn.cache(my_func)`). That is the second overload.

The overloads are vitally important because for user code the most important thing is that, regardless of how `cache` is used, it needs to always return a final function with a type signature that looks the same as the user's original function. This should achieve that.